### PR TITLE
Add configuration methods

### PR DESCRIPTION
**Pull request by juliamagan**
## Description

In this PR, functions have been added that will be used for the configuration of Wazuh nodes. This configuration can be done either in parallel or not.

The following functions have been added which are responsible for modifying and restoring the configuration:

- `configure_host()`.
- `change_agents_configured_manager()`.
- `backup_host_configuration()`.
- `restore_host_backup_configuration()`

In addition, the following functions have been added, which are responsible for executing the above functions in parallel when desired:

- `configure_environment()`.
- `backup_environment_configuration()`.
- `restore_environment_backup_configuration()`.

The configuration expected by these functions will have the following format:

```
wazuh-manager1:
  local_internal_options.conf:
    remoted.debug: '2'
  ossec.conf:
  - section: client
    elements:
    - server:
        elements:
        - address:
            value: 121.1.3.1
  agent.conf:
    group: default
    configuration:
    - section: client
      elements:
      - server:
          elements:
          - address:
              value: 121.1.3.1
  api.yaml:
    logs:
      level: debug

wazuh-agent1:
  ossec.conf:
    ...
```

The supported files are `ossec.conf`, `agent.conf`, `api.yaml` and `agent.conf`

### CODE REVIEW

File: src/wazuh_qa_framework/system/wazuh_handler.py [hunk 1]
```diff
@@ -3,20 +3,51 @@
...
     'linux': '/var/ossec',
-    'windows': 'C:\\Program Files\\ossec-agent',
+    'windows': 'C:/Program Files (x86)/ossec-agent',
     'darwin': '/Library/Ossec'
 }
 
+DEFAULT_TEMPORAL_DIRECTORY = {
+    'linux': '/tmp',
+    'windows': 'C:/Users/qa/AppData/Local/Temp'
+}
```
Comment (Rebits) [inline src/wazuh_qa_framework/system/wazuh_handler.py:24]: These vars should be moved into a common variables module

File: src/wazuh_qa_framework/system/wazuh_handler.py [hunk 2]
```diff
@@ -3,20 +3,51 @@
...
+    return new_configuration
+
+
+conf_functions = {
+    'local_internal_options.conf': configure_local_internal_options,
+    'ossec.conf': configure_ossec_conf,
+    'agent.conf': configure_ossec_conf,
+    'api.yaml': configure_api_yaml
+}
+
```
Comment (Rebits) [inline src/wazuh_qa_framework/system/wazuh_handler.py:50]: We should move all these functions to the configuration module.
Functions to get the directory path should be moved to a separate module too.

File: src/wazuh_qa_framework/system/wazuh_handler.py [hunk 3]
```diff
@@ -340,7 +371,7 @@ def get_ruleset_directory_path(self, host):
 
         return ruleset_directory_path
 
-    def configure_host(self, host, configuration_host):
+    def configure_host(self, host, configuration_file, configuration_values):
```
Comment (Rebits) [inline src/wazuh_qa_framework/system/wazuh_handler.py:374]: Update the examples in the docstring

File: src/wazuh_qa_framework/system/wazuh_handler.py [hunk 4]
```diff
@@ -369,9 +400,36 @@ def configure_host(self, host, configuration_host):
...
+        configuration_values = (configuration_values['configuration'] if configuration_file == 'agent.conf'
+                                else configuration_values)
+
+        # Get configuration file path
+        host_configuration_file_path = self.get_file_fullpath(host, configuration_file, group)
+
+        parameters = {'new_conf': configuration_values}
+
+        # Get template for ossec.conf and agent.conf
+        if configuration_file == 'ossec.conf' or configuration_file == 'agent.conf':
```
Comment (Rebits) [inline src/wazuh_qa_framework/system/wazuh_handler.py:423]:
```suggestion
        if configuration_file in ['ossec.conf', 'agent.conf']:
```
We can also move that list into a variable called xml_configuration_files

**Review by Rebits — CHANGES_REQUESTED** [review]
GJ! Some minor changes are required

File: src/wazuh_qa_framework/system/wazuh_handler.py (hunk 1 above)
Comment (juliamagan) [reply src/wazuh_qa_framework/system/wazuh_handler.py:24]: As the development can be long and goes out of the scope of this PR, the following issue has been created and will be worked on:

https://github.com/wazuh/qa-system-framework/issues/74

File: src/wazuh_qa_framework/system/wazuh_handler.py (hunk 2 above)
Comment (juliamagan) [reply src/wazuh_qa_framework/system/wazuh_handler.py]: Done in ecc6d4c

File: src/wazuh_qa_framework/system/wazuh_handler.py (hunk 3 above)
Comment (juliamagan) [reply src/wazuh_qa_framework/system/wazuh_handler.py:349]: Done in 8aa3f89

File: src/wazuh_qa_framework/system/wazuh_handler.py (hunk 4 above)
Comment (juliamagan) [reply src/wazuh_qa_framework/system/wazuh_handler.py]: Done in 1deaf31

File: src/wazuh_qa_framework/system/wazuh_handler.py [hunk 5]
```diff
@@ -340,108 +346,210 @@ def get_ruleset_directory_path(self, host):
...
         Args:
             configuration_list (dict): Host configuration files to backup
         Returns:
             dict: Host backup filepaths
         """
+        self.logger.debug(f"Creating {file} backup on {host}")
+        backup_paths = {host: {}}
+        host_configuration_file_path = self.get_file_fullpath(host, file, group)
+        temporal_folder = DEFAULT_TEMPORAL_DIRECTORY[self.get_ansible_host_os(host)]
+        backup_file = os.path.join(temporal_folder, file + '.backup',)
```
Comment (BelenValdivia) [inline src/wazuh_qa_framework/system/wazuh_handler.py:479]: There is an extra comma at the end of the `backup_file` definition

File: src/wazuh_qa_framework/system/wazuh_handler.py [hunk 6]
```diff
@@ -340,108 +346,210 @@ def get_ruleset_directory_path(self, host):
...
+        self.copy_file(host=host, dest_path=dest_file,
+                       src_path=backup_file, remote_src=True, become=not self.is_windows(host))
+        self.logger.debug(f"Restored {dest_file} backup on {host} succesfully")
 
-    def restore_environment_backup_configuration(self, backup_configuration, parallel=True):
+    def restore_environment_backup_configuration(self, backup_configurations, parallel=True):
         """Restore environment backup configuration
 
         Args:
             backup_configuration (dict): Backup configuration filepaths
```
Comment (BelenValdivia) [inline src/wazuh_qa_framework/system/wazuh_handler.py:539]: Argument name `backup_configurations (dict):`...

**Review by BelenValdivia — CHANGES_REQUESTED** [review]

File: src/wazuh_qa_framework/system/wazuh_handler.py (hunk 6 above)
Comment (juliamagan) [reply src/wazuh_qa_framework/system/wazuh_handler.py]: Done in 49f9322

File: src/wazuh_qa_framework/system/wazuh_handler.py (hunk 5 above)
Comment (juliamagan) [reply src/wazuh_qa_framework/system/wazuh_handler.py]: Done in 49f9322

File: src/wazuh_qa_framework/system/wazuh_handler.py [hunk 7]
```diff
@@ -340,108 +346,210 @@ def get_ruleset_directory_path(self, host):
...
+                       become=not self.is_windows(host))
+
+        self.logger.debug(f"Created {file} backup on {host} successfully")
+        return backup_paths
+
+    def backup_environment_configuration(self, configuration_hosts, parallel=True):
         """Backup specified files in all hosts
 
         Args:
             configuration_list (dict): Host configuration files to backup
```
[the remark this answers is not in the document]
Comment (juliamagan) [reply src/wazuh_qa_framework/system/wazuh_handler.py]: Done in 49f9322

File: src/wazuh_qa_framework/generic_modules/tools/configuration.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+        """
+        line_counter = 0
+        for line in str_list:
+            line_counter += 1
+            if root_delimeter in line:
+                return str_list[0:line_counter]
+        else:
+            return str_list
+
+    def to_elementTree(str_list: List[str], root_delimeter: str = "</ossec_config>") -> ET.ElementTree:
```
Comment (QU3B1M) [inline src/wazuh_qa_framework/generic_modules/tools/configuration.py:134]: This function's name is in a mix of snake case and camel case. It is recommended to use only snake case in Python.
```suggestion
    def to_element_tree(str_list: List[str], root_delimeter: str = "</ossec_config>") -> ET.ElementTree:
```

Comment (juliamagan) [reply src/wazuh_qa_framework/generic_modules/tools/configuration.py]: Done in 72828b5

**Comment by QU3B1M** [comment]
LGTM! 🚀

**Comment by BelenValdivia** [comment]
LGTM!

**Review by BelenValdivia — APPROVED** [review]

**Pull request merged**